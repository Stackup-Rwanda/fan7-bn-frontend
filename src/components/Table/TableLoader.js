import React from 'react';
import './styles/TableLoader.scss';

const TableLoader = () => (
  <div className="table-ellipsis">
    <div></div>
    <div></div>
    <div></div>
    <div></div>
  </div>
);

export default TableLoader;
