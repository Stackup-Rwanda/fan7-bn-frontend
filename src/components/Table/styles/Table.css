.data-table {
  width: auto;
  padding-top: 1em;
  background: white;
}

.data-table__header {
  border-top: 0.5px solid rgba(211, 210, 210, 0.5);
  border-bottom: 0.5px solid rgba(211, 210, 210, 0.5);
}

.data-table__header label {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
      -ms-flex-direction: row;
          flex-direction: row;
  -webkit-box-pack: end;
      -ms-flex-pack: end;
          justify-content: flex-end;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  padding: 0.5em 1em;
  font-size: 14px;
  font-family: 'Poppins', sans-serif;
}

.data-table__header span {
  color: #555252 !important;
  margin-right: 1em;
}

.data-table__header___search {
  display: block;
  width: auto !important;
  height: 14px !important;
  padding: 6px 12px;
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
  color: #555 !important;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid rgba(0, 0, 0, 0.4) !important;
  border-radius: 0 !important;
  margin-bottom: 0;
}

.data-table__header___search:focus {
  outline: 0;
  -webkit-box-shadow: inset 0 0.5px 0.5px rgba(0, 0, 0, 0.075), 0 0 8px rgba(5, 108, 242, 0.6);
  box-shadow: inset 0 0.5px 0.5px rgba(0, 0, 0, 0.075), 0 0 8px rgba(5, 108, 242, 0.6);
}

.data-table__header___search::-moz-placeholder {
  color: #999;
  opacity: 1;
}

.data-table__header___search:-ms-input-placeholder {
  color: #999;
}

.data-table__header___search::-webkit-input-placeholder {
  color: #999;
}

.data-table table {
  width: 100%;
  text-align: left;
  font-size: 14px;
  padding: 0 .5em;
  border-top: 0.5px solid rgba(211, 210, 210, 0.5);
  border-bottom: 0.5px solid rgba(211, 210, 210, 0.5);
}

.data-table table th,
.data-table table td {
  padding: 10px 5px;
}

.data-table table th {
  border-bottom: 0.2px solid rgba(211, 210, 210, 0.5);
}

.data-table table tr {
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
}

.data-table table tr:nth-child(even) {
  background-color: rgba(236, 236, 236, 0.5);
}

.data-table table td .actionBtn {
  display: block;
  font-weight: normal;
  text-align: center;
  white-space: nowrap;
  background-color: transparent;
  background-image: none;
  -ms-touch-action: manipulation;
  touch-action: manipulation;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: none;
  cursor: pointer;
}

.data-table table td .actionBtn:focus,
.data-table table td .actionBtn:active .data-table table td .actionBtn:active:focus {
  outline: 0;
  background-image: none;
}

.data-table table td img {
  height: 1.5em;
}

.data-table table td .editBtn {
  background-color: #4081d6;
  border-radius: 30px;
  padding: 12px 15px 12px 15px;
  width: 100%;
  font-size: 13px !important;
  color: white !important;
  font-weight: bolder;
  border: 0;
}
/*# sourceMappingURL=Table.css.map */