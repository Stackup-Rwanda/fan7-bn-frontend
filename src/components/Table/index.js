import React from 'react';
import PropTypes from 'prop-types';
import TableLoader from './TableLoader';
import './styles/Table.scss';
import options from '../../assets/icons/icons8-menu-vertical-30.png';

const renderData = (data, cols, actions, handleAction) =>
  data.map((row) => (
    <tr key={row.id}>
      {!!cols &&
        cols.map((col) => (
          <td key={col.name}>
            {Array.isArray(row[col.name])
              ? row[col.name].map((arr) => (
                  <>
                    {arr}
                    <br />
                  </>
                ))
              : row[col.name]}
          </td>
        ))}
      {actions && (
        <td key="actions">
          <div className="dropdown">
            <button
              className="dropbtn"
              type="button"
              id={`table_dots_btn_${row.id}`}
              onClick={(e) => handleAction(row, e)}
            >
              <img src={options} alt="more options" />
            </button>
           </div>
        </td>
      )}
    </tr>
  ));

const renderEmptyState = (cols, actions) => <tr><td colSpan={actions ? cols.length + 1 : cols.length}>There is no data in this table</td></tr>;

const Table = ({ cols, data, loading, actions, handleAction }) => {
  return (
    <div className="data-table">
      <table className="table">
        <thead>
          <tr>
            {cols.map(col => (
              <th key={col.name}>{col.header}</th>
            ))}
            {actions && <th key="actions">Actions</th>}
          </tr>
        </thead>
        <tbody>
          {loading ? (
            <div className="table-loader"><TableLoader /></div>
          ) : data.length > 0 ? (
            renderData(data, cols, actions, handleAction)
          ) : (
            renderEmptyState(cols, actions)
          )}
        </tbody>
      </table>
    </div>
  );
};

Table.propTypes = {
  cols: PropTypes.arrayOf(),
  data: PropTypes.arrayOf(),
  loading: PropTypes.bool,
  actions: PropTypes.bool,
  handleAction: PropTypes.func
};

Table.defaultProps = {
  cols: [
    { header: 'ID', name: 'id' },
    { header: 'Name', name: 'name' },
    { header: 'Email', name: 'email' },
    { header: 'Date', name: 'date' },
  ],
  data: [
    { id: 1, name: 'Elvis', email: 'elvisrugamba@example.com', date: ['1997-10-12', '1997-10-12'] },
  ],
  loading: false,
  actions: false,
  handleAction: () => {}
};

export default Table;