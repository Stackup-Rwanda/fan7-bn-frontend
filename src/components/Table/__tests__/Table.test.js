/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import Table from '../index';

describe('Table Component Test', () => {
  it('should return true if component exists', () => {
    const wrapper = shallow(<Table />);

    expect(wrapper.find('table').exists()).toBe(true);
  });

  it('should have ONLY 1 thead element', () => {
    const wrapper = shallow(<Table />);

    expect(wrapper.find('thead')).toHaveLength(1);
  });

  it('should have ONLY 1 tbody tag', () => {
    const wrapper = shallow(<Table />);

    expect(wrapper.find('tbody')).toHaveLength(1);
  });

  it('should display Loader if data is being fetched', () => {
    const wrapper = shallow(<Table loading={true} />);

    expect(wrapper.find('.table-loader').exists()).toBe(true);
  });

  it('should renders in table rows based on provided columns', () => {
    const cols = [
      { header: 'ID', name: 'id' },
      { header: 'Name', name: 'name' },
      { header: 'Email', name: 'email' },
      { header: 'Language(s)', name: 'languages' },
    ];
    const data = [
      { id: 1, name: 'Elvis', email: 'elvisrugamba@example.com', languages: ['English', 'French'] },
      { id: 2, name: 'Sharon', email: 'sharon@example.com', languages: ['English', 'French'] },
      {
        id: 3,
        name: 'Clement',
        email: 'clement@example.com',
        languages: ['English', 'French'],
        someTest: 10,
      },
    ];

    const wrapper = shallow(<Table data={data} cols={cols} />);

    const table = wrapper.find('table');
    expect(table).toHaveLength(1);

    const thead = table.find('thead');
    expect(thead).toHaveLength(1);

    const headers = thead.find('th');
    expect(headers).toHaveLength(cols.length);
    headers.forEach((th, idx) => {
      expect(th.text()).toEqual(cols[idx].header);
    });

    const tbody = table.find('tbody');
    expect(tbody).toHaveLength(1);

    const rows = tbody.find('tr');
    expect(rows).toHaveLength(data.length);
    rows.forEach((tr, rowIndex) => {
      const cells = tr.find('td');
      expect(cells).toHaveLength(cols.length);
      // expect(cells.at(0).text()).toEqual(data[rowIndex].id);
      expect(cells.at(1).text()).toEqual(data[rowIndex].name);
      expect(cells.at(2).text()).toEqual(data[rowIndex].email);
    });
  });

  it('should display actions butoon if actions is optted in', () => {
    const cols = [
      { header: 'ID', name: 'id' },
      { header: 'Name', name: 'name' },
      { header: 'Email', name: 'email' },
      { header: 'Language(s)', name: 'languages' },
    ];
    const data = [
      { id: 1, name: 'Elvis', email: 'elvisrugamba@example.com', languages: ['English', 'French'] },
      { id: 2, name: 'Sharon', email: 'sharon@example.com', languages: ['English', 'French'] },
      {
        id: 3,
        name: 'Clement',
        email: 'clement@example.com',
        languages: ['English', 'French'],
        someTest: 10,
      },
    ];

    const wrapper = shallow(
      <Table data={data} cols={cols} actions={true} handleAction={() => {}} />
    );

    const table = wrapper.find('table');
    expect(table).toHaveLength(1);

    const thead = table.find('thead');
    expect(thead).toHaveLength(1);

    const headers = thead.find('th');
    expect(headers).toHaveLength(cols.length + 1);

    const tbody = table.find('tbody');
    expect(tbody).toHaveLength(1);

    const rows = tbody.find('tr');
    expect(rows).toHaveLength(data.length);
    rows.forEach((tr, rowIndex) => {
      const cells = tr.find('td');
      expect(cells).toHaveLength(cols.length + 1);
      expect(
        cells
          .last()
          .find('button')
          .exists()
      ).toBe(true);
    });
  });

  it('renders empty message as table cell if there is no data', () => {
    const cols = [
      { header: 'ID', name: 'id' },
      { header: 'Name', name: 'name' },
      { header: 'Email', name: 'email' },
      { header: 'Language(s)', name: 'languages' },
    ];

    const wrapper = shallow(<Table data={[]} cols={cols} />);

    const table = wrapper.find('table');
    expect(table).toHaveLength(1);

    const thead = table.find('thead');
    expect(thead).toHaveLength(1);

    const headers = thead.find('th');
    expect(headers).toHaveLength(cols.length);

    headers.forEach((th, idx) => {
      expect(th.text()).toEqual(cols[idx].header);
    });

    const tbody = table.find('tbody');
    expect(tbody).toHaveLength(1);

    const row = tbody.find('tr');
    expect(row).toHaveLength(1);

    const cell = row.find('td');
    expect(cell).toHaveLength(1);

    expect(cell.prop('colSpan')).toEqual(cols.length);

    expect(cell.text()).toEqual('There is no data in this table');
  });

  it('should renders empty message as table cell if there is no data and colspan to equal cols.length + 1 if actions is opted in', () => {
    const cols = [
      { header: 'ID', name: 'id' },
      { header: 'Name', name: 'name' },
      { header: 'Email', name: 'email' },
      { header: 'Language(s)', name: 'languages' },
    ];

    const wrapper = shallow(<Table data={[]} cols={cols} actions={true} />);

    const table = wrapper.find('table');
    expect(table).toHaveLength(1);

    const thead = table.find('thead');
    expect(thead).toHaveLength(1);

    const headers = thead.find('th');
    expect(headers).toHaveLength(cols.length + 1);

    const tbody = table.find('tbody');
    expect(tbody).toHaveLength(1);

    const row = tbody.find('tr');
    expect(row).toHaveLength(1);

    const cell = row.find('td');
    expect(cell).toHaveLength(1);

    expect(cell.prop('colSpan')).toEqual(cols.length + 1);

    expect(cell.text()).toEqual('There is no data in this table');
  });
});

describe('<Table /> shallow rendering tests', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<Table />);
    expect(tree).toMatchSnapshot();
  });
});
