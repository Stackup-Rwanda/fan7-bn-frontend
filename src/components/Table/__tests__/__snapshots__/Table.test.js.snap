// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> shallow rendering tests matches the snapshot 1`] = `
<div
  className="data-table"
>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          key="id"
        >
          ID
        </th>
        <th
          key="name"
        >
          Name
        </th>
        <th
          key="email"
        >
          Email
        </th>
        <th
          key="date"
        >
          Date
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="1"
      >
        <td
          key="id"
        >
          1
        </td>
        <td
          key="name"
        >
          Elvis
        </td>
        <td
          key="email"
        >
          elvisrugamba@example.com
        </td>
        <td
          key="date"
        >
          1997-10-12
          <br />
          1997-10-12
          <br />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
